_drafts/javascript.md

```diff
@@ -0,0 +1,463 @@
+`require('hubot/es2015')` which returns our es2015 class declarations, so that
+3rd party developers can use es2015 class declarations themselves and inhert
+from ours.
+
+## Detailed process
+
+The steps of the conversation will be
+
+- [x] Convert source files from CoffeeScript to JavaScript.
+- [x] Test with popular projects depending on Hubot
+- [x] Convert test files from CoffeeScript to JavaScript.
+- [x] update package.json
+- [x] add script for JavaScript linting
+- [ ] Release `hubot@3.0.0`
+  - [ ] Update hubot’s version range `peerDependencies` of default scripts
+  - [ ] Publish `hubot@3.0.0` with `--tag=next`
+  - [ ] Convert default scripts to JavaScrip, release as new breaking versions with `@next` tag
+  - [ ] Convert generator to JavaScript, release new breaking version with `@next` tag
+  - [ ] Test all `@next` releases
+  - [ ] Set `@latest` tag on all modules
+- [ ] Update documentation
+
+All steps will be part of the same PR but split into separate commits so people
+can follow more easily.
+
+See also [examples](#examples) below
+
+### Convert source files from CoffeeScript to JavaScript
+
+Convert all source files from CoffeeScript to JavaScript with a tool like [decaffeinate](https://github.com/decaffeinate/decaffeinate), see also [github/hubot#1138](https://github.com/github/hubot/issues/1138).
+
+Make sure the existing tests (still written in CoffeeScript) run against the new JavaScript.
+
+Now go through each file and improve the code readability by hand as needed. From my experiences with converting Hoodie from CoffeeScript to JavaScript quite a lot of manual work will be required.
+
+### Test with popular projects depending on Hubot
+
+- [x] [slack](https://github.com/slackapi/hubot-slack)
+- [x] [hipchat](https://github.com/hipchat/hubot-hipchat)
+- [x] [irc](https://github.com/nandub/hubot-irc)
+
+Instructions to test an existing adapter with the [pull request](https://github.com/github/hubot/pull/1347):
+
+1. Check out the pull request
+
+   ```
+   git clone git@github.com:github/hubot.git
+   cd hubot
+   npm install
+   npm link
+   ```
+
+2. Now link it into your test app using an example adapter
+
+   Using the example of slack here:
+
+   ```
+   npm install -g yo generator-hubot
+   mkdir my-awesome-hubot && cd my-awesome-hubot
+   yo hubot --adapter=slack
+   npm link hubot
+   ```
+
+3. Now start hubot
+
+   ```
+   NODE_PATH=`pwd`/node_modules HUBOT_SLACK_TOKEN=<your token here> ./bin/hubot --adapter slack
+   ```
+
+   replace `<your token here>` with your own bots token from slack
+
+### Convert test files from CoffeeScript to JavaScript
+
+Same as the previous steps only for the tests files this time.
+
+When finished, create a separate branch with the tests written in JavaScript but the source code still in CoffeeScript to assure integrity.
+
+### Update package.json
+
+Remove all CoffeeScript related dependencies and tooling. Update [engines](https://docs.npmjs.com/files/package.json#engines)
```

Comment:
startet PR for engines:
https://github.com/hubotio/hubot/pull/1358

We still have the`coffee-script` dependency because [`bin/hubot` still depends on it](https://github.com/hubotio/hubot/blob/d53e399fa0a6a2f0b441aafb869d1640d12785e0/bin/hubot#L1). We left it in so that existing apps written in CoffeeScript will continue to work.

Is there a better way to achieve the same?